js_modules/dagit/packages/core/src/nav/useCodeLocationsStatus.tsx

```diff
@@ -29,12 +42,37 @@ export const useCodeLocationsStatus = (skip = false): StatusAndMessage | null =>
     fetchPolicy: 'network-only',
     notifyOnNetworkStatusChange: true,
     skip,
+    onCompleted: (data: CodeLocationStatusQuery) => {
+      const entries =
+        data?.locationStatusesOrError?.__typename === 'WorkspaceLocationStatusEntries'
+          ? data?.locationStatusesOrError.entries
+          : [];
+
+      let hasUpdates = entries.length !== Object.keys(state.entriesById || {}).length;
+      const entriesById = {};
+      entries.forEach((entry) => {
+        const previousEntry = state.previousEntriesById && state.previousEntriesById[entry.id];
+        const entryIsUpdated =
+          !previousEntry ||
+          previousEntry.updateTimestamp < entry.updateTimestamp ||
+          previousEntry.loadStatus !== entry.loadStatus;
+        hasUpdates = hasUpdates || entryIsUpdated;
+        entriesById[entry.id] = entryIsUpdated
+          ? {
+              id: entry.id,
+              loadStatus: entry.loadStatus,
+              updateTimestamp: entry.updateTimestamp,
+            }
+          : previousEntry;
+      });
+      if (hasUpdates) {
+        setState({entriesById, previousEntriesById: state.entriesById});
+      }
```

Comment:
Since you're now using `onCompleted`, and since (I think) the effect is just responding to the state change, you might be able to put all of the effect behavior into here and remove the effect entirely.

Also, I think you could put this inside the setter so you don't risk ending up in some weird situation where `state` is some stale value that you don't expect. Not sure how likely that is, though.

```
setState((current) => {
  const {entriesById} = current;
  let hasUpdates = entries.length !== ...;
  // etc.

  return hasUpdates ? {entriesById: newEntriesById, previousEntriesById: entriesById} : current;
});
```